substantial/backends/redis.py

```diff
@@ -11,70 +9,133 @@
 class RedisBackend(Backend):
     def __init__(self, host: str, port: int, **kwargs):
         self.redis = redis.Redis(host=host, port=port, decode_responses=True, **kwargs)
+        self.separator = ":/"
+        self.base_prefix = "substantial"
         # self.redis.flushall()
 
+    # Utils
+    def _key(self, *parts: str) -> str:
+        invalid_chunks = [part for part in parts if self.separator in part]
+        if len(invalid_chunks) > 0:
+            raise ValueError(
+                f"Fatal: parts {invalid_chunks} cannot contain separator '{self.separator}'"
+            )
+        return f"{self.base_prefix}{self.separator}{self.separator.join(parts)}"
+
+    def _parts(self, key: str) -> List[str]:
+        if not key.startswith(self.base_prefix):
+            raise ValueError(
+                f"Invalid key '{key}': required prefix '{self.base_prefix}' not present"
+            )
+
+        return key[len(self.base_prefix) :].split(self.separator)
+
+    # Backend
+
     async def read_events(self, run_id: str) -> Union[Records, None]:
-        key = "_".join(["runs", run_id, "events"])
-        val = self.redis.get(key)
+        event_key = self._key("runs", run_id, "events")
+        val: str | None = self.redis.register_script("""
+            local event_key = KEYS[1]
+            if redis.call("EXISTS", event_key) == 1 then
+                return redis.call("GET", event_key)
+            else
+                return nil
```

Comment:
doesn't get already return nil when the key doesn't exist?

Reply:
Good point, I don't quite remember but I think I had conversion error or something like that from lua world to rust, also here I pretty much wanted to sync both versions as I did some improvements on the [metatype version](https://github.com/metatypedev/metatype/blob/main/src/substantial/src/backends/redis.rs).

Could be an outdated issue though.